### PR TITLE
Add the TestStream Primitive

### PR DESCRIPTION
**Pull request by tgroh**
This is a source suitable for use with tests that have interesting
triggering behavior. It is an Unbounded source that emits elements in
bundles, and advances the watermark and processing time appropriately.

TestStream is currently only supported in the InProcessPipelineRunner.

backports https://github.com/apache/incubator-beam/pull/817
backports https://github.com/apache/incubator-beam/pull/907

### CODE REVIEW

**Comment by tgroh** [comment]
R: @bjchambers

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineOptions.java [hunk 1]
```diff
@@ -67,7 +56,7 @@
 
   void setClock(Clock clock);
 
-  @Default.Boolean(false)
+  @Default.Boolean(true)
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineOptions.java]: Seems like a semantic change for users who weren't setting this -- why is this OK?

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineOptions.java [hunk 2]
```diff
@@ -56,9 +47,7 @@
    * Gets the {@link Clock} used by this pipeline. The clock is used in place of accessing the
    * system time when time values are required by the evaluator.
    */
-  @Default.InstanceFactory(NanosOffsetClock.Factory.class)
   @JsonIgnore
-  @Required
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineOptions.java]: Why is this no longer required? Should it be deprecated?

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineRunner.java [hunk 3]
```diff
@@ -173,13 +179,26 @@
...
+  private Supplier<Clock> clockSupplier;
 
   public static InProcessPipelineRunner fromOptions(PipelineOptions options) {
     return new InProcessPipelineRunner(options.as(InProcessPipelineOptions.class));
   }
 
   private InProcessPipelineRunner(InProcessPipelineOptions options) {
     this.options = options;
+    if (options.getExecutorServiceFactory() == null) {
+      executorServiceSupplier = new FixedThreadPoolSupplier();
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineRunner.java]: Why not use FixedThreadPoolSupplier as the default?

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineRunner.java [hunk 4]
```diff
@@ -173,13 +179,26 @@
...
+    } else {
+      executorServiceSupplier =
+          new ExecutorServiceFactorySupplier(options.getExecutorServiceFactory());
+    }
+    if (options.getClock() == null) {
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineRunner.java]: Why not leave NanosOffsetClock as the default in the options?

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineRunner.java [hunk 5]
```diff
@@ -189,6 +208,14 @@ public InProcessPipelineOptions getPipelineOptions() {
     return options;
   }
 
+  Supplier<Clock> getClockSupplier() {
+    return clockSupplier;
+  }
+
+  void setClockSupplier(Supplier<Clock> supplier) {
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineRunner.java]: I worry about changes in behavior when the clock is changed midway through processing. Why does it need to be mutable?

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/NanosOffsetClock.java [hunk 6]
```diff
@@ -44,15 +40,5 @@ public Instant now() {
         baseMillis + (TimeUnit.MILLISECONDS.convert(
             System.nanoTime() - nanosAtBaseMillis, TimeUnit.NANOSECONDS)));
   }
-
-  /**
-   * Creates instances of {@link NanosOffsetClock}.
-   */
-  public static class Factory implements DefaultValueFactory<Clock> {
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/NanosOffsetClock.java]: This change seems to be noise -- why did this need to change for the TestStream changes?

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/TestStreamEvaluatorFactory.java [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  private <InputT, OutputT> TransformEvaluator<? super InputT> createEvaluator(
+      AppliedPTransform<PBegin, PCollection<OutputT>, TestStream<OutputT>> application,
+      InProcessEvaluationContext evaluationContext) {
+    if (evaluator.get() == null) {
+      Evaluator<OutputT> createdEvaluator = new Evaluator<>(application, evaluationContext, inUse);
+      evaluator.compareAndSet(null, createdEvaluator);
+    }
+    if (inUse.compareAndSet(false, true)) {
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/TestStreamEvaluatorFactory.java]: This logic is somewhat confusing -- can you add javadoc on why you need to do this? At face value, creating an evaluator should be as simple as "return new Evaluator<>(application, evaluationContext, inUse)". I then see that you're caching the evaluator. That makes sense. But I'm not sure why you return null if you've already created one.

(1) Just make which ever thread successfully creates the evaluator be the one to return it?
(2) Or just have a boolean, and only the thread that flips false->true returns the evaluator.

(And also comment why you want to return after the first one)

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/TestStreamEvaluatorFactory.java [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
+
+/**
+ * The {@link TransformEvaluatorFactory} for the {@link TestStream} primitive.
+ */
+class TestStreamEvaluatorFactory implements TransformEvaluatorFactory {
+  private final AtomicBoolean inUse = new AtomicBoolean(false);
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/TestStreamEvaluatorFactory.java]: Javadoc on the purpose of these fields (especially how they should be used safely)?

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/TestStreamEvaluatorFactory.java [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  private static class TestClockSupplier implements Supplier<Clock> {
+    @Override
+    public Clock get() {
+      return new TestClock();
+    }
+  }
+
+  static class DirectTestStreamFactory implements PTransformOverrideFactory {
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/TestStreamEvaluatorFactory.java]: Can the definition of the test version exist in testing code somewhere?

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/TestStreamEvaluatorFactory.java [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+
+      @Override
+      public PCollection<T> apply(PBegin input) {
+        setup(input.getPipeline());
+        return PCollection.<T>createPrimitiveOutputInternal(
+                input.getPipeline(), WindowingStrategy.globalDefault(), IsBounded.UNBOUNDED)
+            .setCoder(original.getValueCoder());
+      }
+
+      private void setup(Pipeline p) {
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/TestStreamEvaluatorFactory.java]: Not entirely clear why this had to be broken out -- were you intending other places to call setup?

Seems like this does both some validation and some mutation. It seems odd that having a DirectTestStream mutates the clock supplier for the whole pipeline. What if you had both a DirectTestStream and other non-DirectTestStream sources?

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/DataflowAssert.java [hunk 11]
```diff
@@ -731,6 +748,12 @@ private PCollectionViewIterableAssert(
     }
 
     @Override
+    public IterableAssert<T> inEarlyGlobalWindowPanes() {
+      throw new UnsupportedOperationException(
+          "IterableAssert of an PCollectionView does not support windowed assertions");
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/DataflowAssert.java]: "of a PCollectionView"

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+import java.io.OutputStream;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * A testing input that generates an unbounded {@link PCollection} of elements, advancing the
+ * watermark and processing time as elements are emitted. After all of the specified elements are
+ * emitted, ceases to produce output.
+ *
+ * <p>Each call to a {@link TestStream.Builder} method will only be reflected in the state of the
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java]: This Javadoc should be replicated on the builder and/or associated methods.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+     */
+    @SafeVarargs
+    public final Builder<T> addElements(
+        TimestampedValue<T> element, TimestampedValue<T>... elements) {
+      checkArgument(
+          element.getTimestamp().isBefore(BoundedWindow.TIMESTAMP_MAX_VALUE),
+          "Elements must have timestamps before %s. Got: %s",
+          BoundedWindow.TIMESTAMP_MAX_VALUE,
+          element.getTimestamp());
+      for (TimestampedValue<T> multiElement : elements) {
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java]: the name "multiElement" makes this confusing, because I thought looking in the code that it was something special, not just one of the elements). I'd suggest renamining the parameters to "firstElement" and "elements" then having the loop be "for element in elements"

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+  /**
+   * The types of {@link Event} that are supported by {@link TestStream}.
+   */
+  public enum EventType {
+    ELEMENT,
+    WATERMARK,
+    PROCESSING_TIME
+  }
+
+  /** A {@link Event} that produces elements. */
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java]: An Event

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+    static <T> Event<T> advanceBy(Duration amount) {
+      return new AutoValue_TestStream_ProcessingTimeEvent<>(EventType.PROCESSING_TIME, amount);
+    }
+  }
+
+  @Override
+  public PCollection<T> apply(PBegin input) {
+    throw new IllegalStateException(
+        String.format(
+            "Pipeline Runner %s does not provide a required override for %s",
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java]: "required override" seems ambiguous (could mean `@Override` method). Maybe just "does not support %s" "does not provide/support an implementation of %s"?

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/testing/TestStreamTest.java [hunk 16]
```diff
@@ -0,0 +1,328 @@
...
+        VarIntCoder.of())
+        .addElements(TimestampedValue.of(1, instant),
+            TimestampedValue.of(2, instant),
+            TimestampedValue.of(3, instant))
+        .advanceWatermarkTo(instant.plus(Duration.standardMinutes(6)))
+        // These elements are late but within the allowed lateness
+        .addElements(TimestampedValue.of(4, instant), TimestampedValue.of(5, instant))
+        .advanceWatermarkTo(instant.plus(Duration.standardMinutes(20)))
+        // These elements are droppably late
+        .addElements(TimestampedValue.of(-1, instant),
```
Comment (bjchambers) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/testing/TestStreamTest.java]: Test would likely be easier to debug if it was separated into different cases -- eg., simplest case, dropping case (just verifying that data was dropped), triggering case, etc.

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/testing/TestStreamTest.java [hunk 17]
```diff
@@ -0,0 +1,328 @@
...
+    p.run();
+  }
+
+  @Test
+  public void testProcessingTimeTrigger() {
+    TestStream<Long> source = TestStream.create(
+        VarLongCoder.of())
+        .addElements(TimestampedValue.of(1L, new Instant(1000L)),
+            TimestampedValue.of(2L, new Instant(2000L)))
+        .advanceProcessingTime(Duration.standardMinutes(12))
```
Comment (bjchambers) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/testing/TestStreamTest.java]: I liked that in the watermark case we're using explicit times rather than durations. Should we instead have `advanceProcessingTimeTo(Instant)` to keep the two consistent (and make testing things like alignedTo and such easier)?

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/testing/TestStreamTest.java [hunk 18]
```diff
@@ -0,0 +1,328 @@
...
+            .advanceWatermarkTo(new Instant(1001L))
+            .addElements(
+                TimestampedValue.of("finalLatePane", new Instant(750)),
+                TimestampedValue.of("alsoFinalLatePane", new Instant(250)))
+            .advanceWatermarkToInfinity();
+
+    Pipeline p = getPipeline();
+    FixedWindows windowFn = FixedWindows.of(Duration.millis(1000L));
+    Duration allowedLateness = Duration.millis(5000L);
+    PCollection<String> values =
```
Comment (bjchambers) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/testing/TestStreamTest.java]: A comment on why the given pipeline actually tests discarding mode. Specifically, here we're testing discarding mode because we trigger multiple times, and if we were in accumulating mode, the elements in the first trigger firing would be duplicated.

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/testing/TestStreamTest.java [hunk 19]
```diff
@@ -0,0 +1,328 @@
...
+  public void testElementAtPositiveInfinityThrows() {
+    Builder<Integer> stream =
+        TestStream.create(VarIntCoder.of())
+            .addElements(TimestampedValue.of(-1, BoundedWindow.TIMESTAMP_MAX_VALUE.minus(1L)));
+    thrown.expect(IllegalArgumentException.class);
+    stream.addElements(TimestampedValue.of(1, BoundedWindow.TIMESTAMP_MAX_VALUE));
+  }
+
+  @Test
+  public void testAdvanceWatermarkNonMonotonicThrows() {
```
Comment (bjchambers) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/testing/TestStreamTest.java]: Add a test that you can start with negative watermarks?

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineOptions.java (hunk 1 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineOptions.java]: Reverted, but set in the TestStream override

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineOptions.java (hunk 2 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineOptions.java]: Yes.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineRunner.java (hunk 3 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineRunner.java]: I have to set it somewhere during construction; it seems reasonable to only create the supplier that will be used.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineRunner.java (hunk 5 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineRunner.java]: The runner will obtain an instance of the Clock during the call to `run()` and pass it to the `EvaluationContext`; so each is independent.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/NanosOffsetClock.java (hunk 6 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/NanosOffsetClock.java]: I needed to replace the way clocks are obtained so multiple pipelines don't interact with each other if they're both using `TestStream`; that requires suppliers, which means we can have a default supplier instead of a DefaultValueFactory.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/TestStreamEvaluatorFactory.java (hunk 7 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/TestStreamEvaluatorFactory.java]: Race condition averted. This should be keyed by application.

Evaluators are stateful (so we can produce the correct outputs) so they have to be cached, but can't be split or parallelized otherwise we get output faster than is expected.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/TestStreamEvaluatorFactory.java (hunk 8 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/TestStreamEvaluatorFactory.java]: Doc'd the one remaining field

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/TestStreamEvaluatorFactory.java (hunk 9 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/TestStreamEvaluatorFactory.java]: I don't understand what this means. This is the runner-specific override.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/TestStreamEvaluatorFactory.java (hunk 10 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/TestStreamEvaluatorFactory.java]: inlined.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/DataflowAssert.java (hunk 11 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/DataflowAssert.java]: Fixed everywhere.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java (hunk 12 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java]: Isn't it? Every builder method has "like this one that will (do a thing) after all earlier events have completed"

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java (hunk 13 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java]: first & rest
for element in rest

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java (hunk 14 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java]: Done.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java (hunk 15 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java]: "does not support" seems fine

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/testing/TestStreamTest.java (hunk 17 above)
Comment (tgroh) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/testing/TestStreamTest.java]: Elements are output with an absolute timestamp, and the watermark is relative to these timestamps. However, Processing Time is always relative to whenever the pipeline happens to be executing, so users should not think of absolute processing times.

If in the future we decide it's important to have both absolute and relative references to processing time, we can revisit this.

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/testing/TestStreamTest.java (hunk 18 above)
Comment (tgroh) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/testing/TestStreamTest.java]: Done.

File: sdk/src/test/java/com/google/cloud/dataflow/sdk/testing/TestStreamTest.java (hunk 19 above)
Comment (tgroh) [inline sdk/src/test/java/com/google/cloud/dataflow/sdk/testing/TestStreamTest.java]: Updated this one.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineRunner.java (hunk 4 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineRunner.java]: Symmetry, and the removal in Beam made this an easier backport

**Comment by tgroh** [comment]
Merged in changes from incubator head, namely https://github.com/apache/incubator-beam/pull/907

PTAL

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InMemoryWatermarkManager.java [hunk 20]
```diff
@@ -1133,19 +1130,12 @@ private void addPending(CommittedBundle<?> bundle) {
...
-        processingTimers = synchronizedProcessingInputWatermark.extractFiredDomainTimers(
-            TimeDomain.PROCESSING_TIME, clock.now());
-        synchronizedTimers = synchronizedProcessingInputWatermark.extractFiredDomainTimers(
-            TimeDomain.SYNCHRONIZED_PROCESSING_TIME, getSynchronizedProcessingInputTime());
-      }
+      Map<StructuralKey<?>, List<TimerData>> processingTimers =
+          synchronizedProcessingInputWatermark.extractFiredDomainTimers(TimeDomain.PROCESSING_TIME,
+              clock.now());
+      Map<StructuralKey<?>, List<TimerData>> synchronizedTimers =
+          synchronizedProcessingInputWatermark.extractFiredDomainTimers(
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InMemoryWatermarkManager.java]: It seems weird that both of these methods are on `synchronizedProcessingInputWatermark`. I see the `TimeDomain` parameter being used to disambiguate, which I guess makes sense. Just wanted to confirm.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineOptions.java [hunk 21]
```diff
@@ -56,9 +47,8 @@
    * Gets the {@link Clock} used by this pipeline. The clock is used in place of accessing the
    * system time when time values are required by the evaluator.
    */
-  @Default.InstanceFactory(NanosOffsetClock.Factory.class)
+  @Deprecated
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineOptions.java]: Add Javadoc on why this is deprecated/what to do instead.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java [hunk 22]
```diff
@@ -0,0 +1,366 @@
...
+      for (int i = 0; i < rest.length; i++) {
+        remainingElements[i] = TimestampedValue.of(rest[i], currentWatermark);
+      }
+      return addElements(firstElement, remainingElements);
+    }
+
+    /**
+     * Adds the specified elements to the source with the provided timestamps.
+     *
+     * @return A {@link TestStream.Builder} like this one that will add the provided elements
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java]: Should "add" be "emit"? The method is adding elements, the stream will cause those elements to be emitted after all earlier events have completed

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java [hunk 23]
```diff
@@ -0,0 +1,366 @@
...
+    private Builder(Coder<T> coder, ImmutableList<Event<T>> events, Instant currentWatermark) {
+      this.coder = coder;
+      this.events = events;
+      this.currentWatermark = currentWatermark;
+    }
+
+    /**
+     * Adds the specified elements to the source with timestamp equal to the current watermark.
+     *
+     * @return A {@link TestStream.Builder} like this one that will add the provided elements
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java]: add -> emit?

**Comment by bjchambers** [comment]
Nits then LGTM

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InMemoryWatermarkManager.java (hunk 20 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InMemoryWatermarkManager.java]: Yes. Processing Time proper doesn't have a meaningful watermark, but processing time timers impact the SynchronizedProcessingTimeWatermark, so they're tracked together.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java (hunk 23 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java]: Sure.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java (hunk 22 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java]: Sure.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineOptions.java (hunk 21 above)
Comment (tgroh) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessPipelineOptions.java]: Done.
